### ISSUE TITLE
Ability to tune the underlying bayes_opt params

### ISSUE THREAD

**Issue by robert1826** [facebookresearch/nevergrad#189]
Some problems require tuning the underlying bayes_opt params such as the utility function being used or even the underlying gp params ... it seems that there is no way to change them using nevergrad

**Comment by jrapin** [comment]
Indeed, the current interface of on optimizer is not very flexible in this regard.
I am trying to push some sort of factory / family of optimizers which could be use for this though. It is mostly hidden for now but some algorithms, including the `BO` variants have one:
https://github.com/facebookresearch/nevergrad/blob/master/nevergrad/optimization/optimizerlib.py#L1044-L1067
In this class, we could add close to anything (though the more we add, the more likely there will be new issues so I try to be cautious about it). Is there anything in particular you would like to have access to?

**Comment by robert1826** [comment]
I would like to have access to the utility function args (kind, kappa and xi) ... and for the GP, I think the alpha parameter will suffice.

**Comment by robert1826** [comment]
another (maybe related) issue is the seed of the BayesianOptimization object ... as I noticed that it's random but I think it should be reproducible / controllable

**Comment by lazyoracle** [comment]
> 
> 
> another (maybe related) issue is the seed of the BayesianOptimization object ... as I noticed that it's random but I think it should be reproducible / controllable

The results of BO can be repeatedly reproduced inside Nevergrad by setting the seed to numpy random as such:
```
import numpy as np
np.random.seed(any_seed)
```

However there are some reported issues with the inability to reproduce BO results across different systems, you might want to take a look at this:
[Different result on different machines for same input and random seed #155](https://github.com/fmfn/BayesianOptimization/issues/155)

**Comment by robert1826** [comment]
@lazyoracle I think you are missing this : https://github.com/facebookresearch/nevergrad/blob/d5bda3886d77e8166118ab46c511addc9855ef83/nevergrad/optimization/optimizerlib.py#L994

**Comment by lazyoracle** [comment]
@robert1826 Won't setting a fixed seed in `np.random()` generate the same `randint()` to be used as the `seed` you mentioned above thus ensuring reproducibility?

**Comment by jrapin** [comment]
@robert1826, as @lazyoracle mentioned, this line is precisely what makes seeding before initialization of the optimizer work. They are unit tests that make sure the results are reproducible (though I did notice slight differences from a computer to another one, which seems consistent with the issue mentioned by @lazyoracle)

For the future, I'm still considering changing how seeding works in `nevergrad` though, providing a default random state that optimizers should pull from.

**Comment by jrapin** [comment]
Concerning the parameters, I'll try to add them next week.

**Comment by jrapin** [comment]
Above is the diff to provide access to more parameters, @robert1826 would that work for you? Can you give more details to what was useful for you and why? This is just for my own culture, I am not yet very familiar with BO.

**Comment by robert1826** [comment]
This looks perfect to me and it's exactly what I asked for .... but for completeness please and another param for the 'init' budget .. instead of just setting it as the sqrt(budget) here:
https://github.com/facebookresearch/nevergrad/blob/b8fdfa72ce6674ff10c9d668d252c9257025c742/nevergrad/optimization/optimizerlib.py#L1004

**Comment by jrapin** [comment]
Ok, this is now changed as well. I am a bit worried that this will create a lot of border cases though, but nevermind for now. On the longer term though, this may mean that the interface will evolve. I see a couple of things that will probably change:
- the first 3 parameters (about the initialization) could(/should?) probably be changed to a unique parameter `init_optimizer` for instance.
- the gaussian cumulative density transform which is used to convert from [0, 1] to R saturates too fast in my experience, I doubt you can get any value beyond 14 or something like this. We may change it to arctan if this makes sense as well.

In a nutshell, don't be surprise if you see some changes (or don't hesitate to ask for more changes, or propose some), this is still under development (sorry for the bother...)

**Comment by jrapin** [comment]
I'll merge now, but am unable to comprehensively test that all the parameters work as intended. Hopefully all should go well but let me know if you see anything weird ;)

**Comment by lazyoracle** [comment]
How well does this integrate with the family/factory of optimizers plan you were mentioning? Do you plan to have a `ParametrizedOptimizer`-like parameter for initialising all the optimizers?

**Comment by robert1826** [comment]
My logic is that the proposed changes are just a way to 'expose' the params of bayes_opt to the user of nevergrad ... so any border cases in this change is probably handled by the bayes_opt package ... anyway I'm an active user and I'll try to test and fix bugs if I find some.

**Comment by jrapin** [comment]
@robert1826 some of the initialization part is not strictly from `bayes_opt` so I won't be able to blame them if it fails :D
@lazyoracle for now there is no clear path, it's more of a tentative API. I will push it whenever this makes sense (it removed a looot of duplicated code one `OnePlusOne` for instance), there is issue #100 to track this process. Still, not all optimizers require parametrization (and not all users want to deal with parametrization) so both optimizers and optimizer family may live together if nobody complains too much...
